### ISSUE TITLE
Custom media dimensions or printable area for Dymo LabelWriter

### ISSUE THREAD

**Issue by jashephe** [michaelrsweet/lprint#19]
I'm trying to print from a Dymo LabelWriter 450 Twin Turbo to [these die-cut labels](https://www.labtag.com/shop/product/cryogenic-labels-for-dymo-1-04-x-0-5-0-375-circle-ec1f-eg1f-081-2/). From the template for Dymo's software provided by the manufacturer (reproduced in [this gist](https://gist.github.com/jashephe/aca05580b7cf211f8ac7b84775eaa4c7)), these labels are derived from [Dymo's #30299 price tag label](https://www.dymo.com/en-US/lw-price-tag-labels--2-up-3-8-x-3-4), which seems to correspond to `oe_price-tag-label_0.9375x0.875in`.

If I attempt to print the test pattern below with `lprint -o media=oe_price-tag-label_0.9375x0.875in -o media-source=alternate-roll labtag081_testpattern.png`, the printed area does not span the full width of the label:

---

Input:
![labtag081_testpattern](https://user-images.githubusercontent.com/936132/77813297-8b5f8900-7075-11ea-8160-6f5d1b70df3b.png)

Result:
![print_result](https://user-images.githubusercontent.com/936132/77813353-f9a44b80-7075-11ea-8fbe-a8bf3fdf97c4.jpg)

---

If I add empty padding on the left side of the test pattern, it just shrinks the printed size, it doesn't shift it to the right. Is there a way to re-define the "printable area" of a label, or define custom margins or label types?

Thank you!

**Comment by michaelrsweet** [comment]
For these sorts of composite/combination labels you need to use a size that covers the full width and height of the label area. So maybe use oe_custom_1.75x0.875in and then provide an image that is padded accordingly.

I can also look at adding support for using the media-xxx-margin values to adjust the position of the rendered image within the label area.

**Comment by jashephe** [comment]
Thanks so much for these suggestions! I tried using `oe_custom_1.75x0.875in` but got an error:
```
$ lprint -o media=oe_custom_1.75x0.875in -o media-source=alternate-roll labtag081_testpattern.png
lprint: Unable to print 'labtag081_testpattern.png' - Unsupported media-col collection value.
```
I did discover that I could specify the dimensions directly with e.g. `media=1.75x0.875in`, but I'm not sure how these dimensions relate to the physical dimensions of the printed label (see below for some tests). Initially, the orientation of the dimensions appeared to be wrong (landscape instead of portrait), and multiple labels were printed simultaneously. Even when I switched from landscape to portrait output, multiple labels were still printed, even when I used a short-side dimension of 0.5" which is measurably narrower than the short side of the label.

---

Both of the following gave the label shown below:
```
$ lprint -o media=1.5x0.875in -o orientation-requested=landscape -o media-source=alternate-roll labtag081_testpattern.png
$ lprint -o media=0.875x1.5in -o orientation-requested=landscape -o media-source=alternate-roll labtag081_testpattern.png
```
![landscape](https://user-images.githubusercontent.com/936132/77826304-97cafc80-70dc-11ea-8e4c-4b5bca923b33.jpg)
(all three labels printed at once)

---

Both of the following gave the label shown below:
```
$ lprint -o media=1.5x0.875in -o orientation-requested=portrait -o media-source=alternate-roll labtag081_testpattern.png
$ lprint -o media=0.875x1.5in -o orientation-requested=portrait -o media-source=alternate-roll labtag081_testpattern.png
```
![portrait](https://user-images.githubusercontent.com/936132/77826310-9f8aa100-70dc-11ea-9b1c-2aaef30427e5.jpg)
(both labels printed at once)

---

I couldn't get it to just print just a single label, even when I went down to e.g.
```
lprint -o media=1.5x0.5in -o orientation-requested=portrait -o media-source=alternate-roll labtag081_testpattern.png
```
![smaller](https://user-images.githubusercontent.com/936132/77826315-a44f5500-70dc-11ea-9f04-4d43289d2d91.jpg)
(both labels printed at once)

**Comment by michaelrsweet** [comment]
Try measuring the width of the carrier backing and the distance between the holes - use that as the dimension of the label. (I was estimating 1.75x0.875 based on the description on the web site you provided...)

**Comment by jashephe** [comment]
I've just gone ahead and measured to confirm, but those dimensions are more or less correct. I measure 1.875 in as the width of the carrier backing (from the left side of the roll to the right side of the roll), and 0.875 in between holes.

**Comment by michaelrsweet** [comment]
Hmm, I'll see if I can find anything applicable in the documentation I have...

**Comment by jashephe** [comment]
Thank you! Interestingly, no matter if I give the dimensions as 1.875x0.875in or 0.875x1.875in, the width of the printed area seems to more or less stay the same — see below for more examples.

```
$ lprint -o media=1.875x0.875in -o orientation-requested=portrait -o media-source=alternate-roll labtag081_testpattern.png
$ lprint -o media=0.875x1.875in -o orientation-requested=portrait -o media-source=alternate-roll labtag081_testpattern.png
```
![portrait2](https://user-images.githubusercontent.com/936132/77827633-d5cc1e80-70e4-11ea-8e28-190af514b01f.jpg)

```
$ lprint -o media=1.875x0.875in -o orientation-requested=landscape -o media-source=alternate-roll labtag081_testpattern.png
$ lprint -o media=0.875x1.875in -o orientation-requested=landscape -o media-source=alternate-roll labtag081_testpattern.png
```
![landscape2](https://user-images.githubusercontent.com/936132/77827643-e2507700-70e4-11ea-930b-bdf871bfd430.jpg)

And again, thank you so much for already being so helpful; sorry to keep coming back with more issues.

**Comment by jashephe** [comment]
Just wanted to share an update, if it helps narrow things down — it appears to just not be honoring the custom label dimensions. When I tried both `-o media=0.1x0.1in -o orientation-requested=portrait` and `-o media=10x10in -o orientation-requested=portrait`, the output was exactly the same as the image I posted previously for `-o media=1.875x0.875in -o orientation-requested=portrait` (and the same is true in the landscape orientation).

The label dimensions with which it is printing seem to match the default label `oe_address-label_1.25x3.5in`.

**Comment by michaelrsweet** [comment]
@jashephe I know it's been a long time, but the current code is now based on PAPPL and should be much more robust. But in particular you should configure the default media size to the dimensions of the whole label area (relatively easy to do through the web interface...)

**Comment by jashephe** [comment]
Hi Michael, thank you for following up! I do still have that printer and those labels in my possession, so I will give it another go and report back here (although it may take me a few weeks to get around to it). Thanks again, really!

**Comment by jashephe** [comment]
Hi Michael, even with fiddling around in the web interface, I was not able to get this to work. I also encountered one or two `SIGSEGV` crashes of the server while using the web interface, one for instance after clicking the "Print Test Page" button. Eventually, I got things into a state where I couldn't get the printer to print at all (although the jobs were reported in the web interface as "Completed"). I'm sure that mis-configuration on my part played a role, as I had limited time to play around. At this point, I don't have the time for detailed investigation, so I'm just going to close this issue so it isn't littering the tracker. If I have the chance to revisit this and encounter problems in a more formal attempt, I'll open new issues. Thank you again for all of your assistance, and for remaining so responsive!
